queries.sql

```diff
@@ -7,4 +7,47 @@ SELECT date_of_birth FROM animals WHERE name IN ('Agumon', 'Pikachu');
 SELECT name, escape_attempts FROM animals WHERE weight_kg > 10.5;
 SELECT * FROM animals WHERE neutered = B'1';
 SELECT * FROM animals WHERE name <> 'Gabumon';
-SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
\ No newline at end of file
+SELECT * FROM animals WHERE weight_kg BETWEEN 10.4 AND 17.3;
+
+/* Add unspecified to species in using transaction and rollback */
+
+BEGIN;
+UPDATE animals SET species = 'unspecified';
+SELECT * FROM animals;
+ROLLBACK;
+SELECT * FROM animals;
+
+/* Add pokemon and digimon to species */
+
+BEGIN;
+UPDATE animals SET species = 'digimon' WHERE name LIKE '%mon';
+UPDATE animals SET species = 'pokemon' WHERE species IS NULL;
+Select * from animals;
+COMMIT;
+Select * from animals;
+
+/* Delete all records */
+
+BEGIN;
+DELETE FROM animals;
+ROLLBACK;
+SELECT * FROM animals;
+
+/* Update weight_kg column and delete animal */
+
+BEGIN;
+DELETE FROM animals WHERE date_of_birth > '2022-01-01';
+SAVEPOINT last_savepoint;
+UPDATE animals SET weight_kg = weight_kg * -1;
+ROLLBACK TO last_savepoint;
+UPDATE animals SET weight_kg = weight_kg * -1 WHERE weight_kg < 0;
+COMMIT;
+
+/* */
+
+SELECT COUNT(*) as animals_count FROM animals;
+SELECT COUNT(*) as never_escape FROM animals WHERE escape_attempts = 0;
+SELECT AVG(weight_kg) FROM animals;
+SELECT neutered, SUM(escape_attempts) as escape_times FROM animals GROUP BY neutered ORDER BY escape_times DESC;
+SELECT species, MIN(weight_kg) as min_weight, MAX(weight_kg) as max_weight FROM animals GROUP BY species;
+SELECT species, AVG(escape_attempts) as avg_escape_attempts FROM animals WHERE date_of_birth BETWEEN '1990-01-01' AND '2000-12-31' GROUP BY species;
```

Comment:
## [OPTIONAL]

When writing SQL queries, it's important to follow a specific order of clauses to ensure the query is structured correctly and produces the desired results. The typical order of writing SQL queries is as follows:

1. SELECT: Begin by specifying the columns or expressions you want to retrieve from the database. Use the SELECT clause followed by a comma-separated list of column names or expressions. For example: SELECT column1, column2 FROM table_name.

2. FROM: Identify the table or tables from which you want to retrieve the data. Use the FROM clause followed by the table name or multiple table names if you need to join tables. For example: FROM table_name.

3. WHERE: Apply conditions to filter the data based on specific criteria. Use the WHERE clause followed by the conditions that need to be met. For example WHERE column_name = value.

4. GROUP BY: If you want to group the data based on specific columns, use the GROUP BY clause followed by the column names. For example GROUP BY column_name.

5. HAVING: Apply conditions to the grouped data using the HAVING clause. It functions similarly to the WHERE clause but operates on the grouped data. For example: HAVING aggregate_function(column_name) > value.

6. ORDER BY: Specify the order in which you want the data to be sorted. Use the ORDER BY clause followed by the column name(s) and the desired sort order (ASC for ascending or DESC for descending). For example: ORDER BY column_name DESC.

7. LIMIT/OFFSET: If you want to limit the number of rows returned or skip a certain number of rows, use the LIMIT and OFFSET clauses. For example: LIMIT 10 OFFSET 5 will return 10 rows starting from the 6th row.

It's important to note that not all clauses are required in every query. The order and presence of clauses may vary based on the specific requirements of your query. Also, keep in mind that different database management systems may have slight variations in syntax or additional clauses.

By following the recommended order of writing SQL queries, you can structure your queries in a logical and effective manner, ensuring accurate and efficient data retrieval.